### ISSUE TITLE
osqueryi returns varrying results when querying routes on the CLI

### ISSUE THREAD

**Issue by dennisjac** [facebook/osquery#2591]
At least on CentOS 7 it seems there is a difference when providing a query to osqueryi on the command line and entering it interactively:

```
[root@virt86 ~]# osqueryi --json "select * from routes where destination = '::1'"
[

]
[root@virt86 ~]# osqueryi
osquery - being built, with love, at Facebook

Using a virtual database. Need help, type '.help'
osquery> select * from routes where destination = '::1';
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| destination | netmask | gateway | source | flags | interface | mtu | metric | type  |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| ::1         | 0       |         |        | 0     | lo        | 0   | 256    | local |
| ::1         | 0       |         |        | 0     | lo        | 0   | 0      | local |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
```

As you can see when proving the query on the command line osqueryi returns no routes yet when entered interactively route are returned.

**Comment by theopolis** [comment]
```
[teddy] /tmp » osqueryi --version
osqueryi version 2.0.0
------------------------------------------------------------
[teddy] /tmp » osqueryi --json "select * from routes where destination = '::1'"
[
  {"destination":"::1","flags":"0","gateway":"","interface":"lo","metric":"0","mtu":"0","netmask":"0","source":"","type":"local"}
]
------------------------------------------------------------
[teddy] /tmp » osqueryi
osquery - being built, with love, at Facebook
Using a virtual database. Need help, type '.help'
osquery> select * from routes where destination = '::1';
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| destination | netmask | gateway | source | flags | interface | mtu | metric | type  |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| ::1         | 0       |         |        | 0     | lo        | 0   | 0      | local |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
osquery>
```

I cannot reproduce this behavior. Can you include your OS version, osquery version, how you installed osquery, and any other steps to help me triage/debug?

**Comment by dennisjac** [comment]
I'm using the projects yum repository and the latest rpm:
```
[root@virt86 ~]# rpm -qa|grep osquery
osquery-s3-centos7-repo-1-0.0.noarch
osquery-2.0.0-1.el7.x86_64
[root@virt86 ~]# cat /etc/redhat-release 
CentOS Linux release 7.1.1503 (Core) 
[root@virt86 ~]# osqueryi --version
osqueryi version 2.0.0
```
I installed the repo rpm and then the osquery rpm and did no further configuration so far.

**Comment by dennisjac** [comment]
After a bit of experimenting I found something interesting. Apparently this issue only occurs when the ".osquery" directory exists in the home directory (which apparently is automatically created when running osqueryi):
```
[root@virt86 ~]# osqueryi "select * from routes where destination = '::1'"
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| destination | netmask | gateway | source | flags | interface | mtu | metric | type  |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| ::1         | 0       |         |        | 0     | lo        | 0   | 256    | local |
| ::1         | 0       |         |        | 0     | lo        | 0   | 0      | local |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
[root@virt86 ~]# osqueryi "select * from routes where destination = '::1'"
[root@virt86 ~]# rm -rf ~/.osquery/
[root@virt86 ~]# osqueryi "select * from routes where destination = '::1'"
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| destination | netmask | gateway | source | flags | interface | mtu | metric | type  |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
| ::1         | 0       |         |        | 0     | lo        | 0   | 256    | local |
| ::1         | 0       |         |        | 0     | lo        | 0   | 0      | local |
+-------------+---------+---------+--------+-------+-----------+-----+--------+-------+
```
Notice how when first run everything is find but the second run suddenly doesn't output anything.
After removing the ~/.osquery directory osqueryi output is correct again.

**Comment by theopolis** [comment]
I was able to reproduce very briefly, meaning the first time I ran osquery (version 2.0.0) on `CentOS Linux release 7.2.1511 (Core)` with:
```
osqueryi --json "select * from routes where destination = '::1'"
```
It did not return results, however after running again it did. I tried relogging into the machine but now I always see results. Perhaps this is not related to osquery, but rather the first time we call the Netlink APIs to inspect routes?

**Comment by theopolis** [comment]
Let's switch to testing with a more deterministic query, like:
```
osqueryi --json "select * from osquery_info"
```

Or `interface_details`, `system_info`, `rpm_packages`. Maybe this is isolated to route inspection?

**Comment by dennisjac** [comment]
```osqueryi --json "select * from osquery_info"``` works everytime for me but the route query only after I remove ~/.osquery.

**Comment by theopolis** [comment]
This may be another issue coming from our use of Netlink, see https://github.com/facebook/osquery/issues/1660 for a potentially-related issue from last year. Querying the Netlink API seems unreliable and racing against the amount of time our process has been executing.

**Comment by theopolis** [comment]
Removing `~/.osquery` may be a red herring as the first-run takes a bit more IO to create/initialize the database and osquery shell history.

**Comment by dennisjac** [comment]
The .osquery is completely empty when I delete it so if it is really a race then initializing database and history shouldn't come into play.
I wrote a quick test script that executes osqueryi 100 times and also dropped the page, dentry and inode caches and executed a "grep -nri <random string> /usr" on the system to create some IO pressure. Without an rmdir ~/.osquery the script reports no output 100% of the time and with the rmdir it sees output 100% of the time.
While this doesn't completely disprove that #1660 might be the cause I wouldn't completely discount the possibility that something else might be the cause of this.

**Comment by theopolis** [comment]
Good info, thanks for the help @dennisjac! Do you have time to dig into the Linux `routes` table implementation and write some test cases to identify the root cause? It could be as easy as checking some return codes and implementing 1 or 2 retries.

**Comment by theopolis** [comment]
Fixed in https://github.com/facebook/osquery/pull/2684.
